go/vt/vtgate/vstream_manager.go

```diff
@@ -238,33 +293,117 @@ func (vs *vstream) streamFromTablet(ctx context.Context, sgtid *binlogdatapb.Sha
+		for j, event := range events {
+			if event.Type == binlogdatapb.VEventType_GTID {
 				// Update the VGtid and send that instead.
-				sgtid.Gtid = ev.Gtid
-				ev.Type = binlogdatapb.VEventType_VGTID
-				ev.Gtid = ""
-				ev.Vgtid = proto.Clone(vs.vgtid).(*binlogdatapb.VGtid)
-			case binlogdatapb.VEventType_FIELD:
-				ev.FieldEvent.TableName = sgtid.Keyspace + "." + ev.FieldEvent.TableName
-			case binlogdatapb.VEventType_ROW:
-				ev.RowEvent.TableName = sgtid.Keyspace + "." + ev.RowEvent.TableName
+				sgtid.Gtid = event.Gtid
+				events[j] = &binlogdatapb.VEvent{
+					Type:  binlogdatapb.VEventType_VGTID,
+					Vgtid: proto.Clone(vs.vgtid).(*binlogdatapb.VGtid),
+				}
 			}
 		}
-		if err := vs.send(evs); err != nil {
+		if err := vs.send(events); err != nil {
 			return err
 		}
 	}
 	return nil
 }
+
+func (vs *vstream) getJournalEvent(ctx context.Context, sgtid *binlogdatapb.ShardGtid, journal *binlogdatapb.Journal) (*journalEvent, error) {
+	if journal.MigrationType == binlogdatapb.MigrationType_TABLES {
+		return nil, nil
+	}
+
+	vs.mu.Lock()
+	defer vs.mu.Unlock()
+
+	je, ok := vs.journaler[journal.Id]
+	if !ok {
+		log.Infof("Journal encountered: %v", journal)
+		// Identify the list of ShardGtids that match the participants of the journal.
+		je = &journalEvent{
+			journal:      journal,
+			participants: make(map[*binlogdatapb.ShardGtid]bool),
+			done:         make(chan struct{}),
+		}
+		const (
+			undecided = iota
+			matchAll
+			matchNone
+		)
+		mode := undecided
+	nextParticipant:
+		for _, jks := range journal.Participants {
+			for _, inner := range vs.vgtid.ShardGtids {
+				if inner.Keyspace == jks.Keyspace && inner.Shard == jks.Shard {
+					switch mode {
+					case undecided, matchAll:
+						mode = matchAll
+						je.participants[inner] = false
+					case matchNone:
+						return nil, fmt.Errorf("not all journaling participants are in the stream: journal: %v, stream: %v", journal.Participants, vs.vgtid.ShardGtids)
+					}
+					continue nextParticipant
+				}
+			}
+			switch mode {
+			case undecided, matchNone:
+				mode = matchNone
+			case matchAll:
+				return nil, fmt.Errorf("not all journaling participants are in the stream: journal: %v, stream: %v", journal.Participants, vs.vgtid.ShardGtids)
+			}
+		}
+		if mode == matchNone {
+			// Unreachable. Journal events are only added to participants.
+			// But if we do receive such an event, the right action will be to ignore it.
+			return nil, nil
+		}
+		vs.journaler[journal.Id] = je
+	}
+
+	if _, ok := je.participants[sgtid]; !ok {
+		// Unreachable. See above.
+		return nil, nil
+	}
+	je.participants[sgtid] = true
+
+	for _, waiting := range je.participants {
+		if !waiting {
+			// Some participants are yet to join the wait.
+			return je, nil
+		}
+	}
+	// All participants are waiting. Replace old shard gtids with new ones.
+	newsgtids := make([]*binlogdatapb.ShardGtid, 0, len(vs.vgtid.ShardGtids)-len(je.participants)+len(je.journal.ShardGtids))
```

Comment:
should add a comment here and this block of code will be executed at most(or is it exactly?) once per event.